### PR TITLE
[JENKINS-42971] GitSCMFileSystem consumes environment and expands branch name

### PR DESCRIPTION
**Pull request by MartinKosicky**
## [JENKINS-42971](https://issues.jenkins.io/browse/JENKINS-42971) Expand environment for branch names

GitSCMFileSystem now implements new SCM Builder APi that receives the Run<?,?> that can is used to expand branch name with environment variables.

### CODE REVIEW

**Comment by MartinKosicky** [comment]
Hmm i wonder what the "/* Unreliable on Windows and not a platform specific test */" means in the failing test...

**Comment by MarkEWaite** [comment]
> Hmm i wonder what the "/* Unreliable on Windows and not a platform specific test */" means in the failing test...

:) It means that the test has been unreliable in the past on Windows and the maintainers did not have time to understand why it was unreliable on Windows.  History of that is captured in the commit messages in case you want to see the sordid details.

**Comment by MartinKosicky** [comment]
> :) It means that the test has been unreliable in the past on Windows and the maintainers did not have time to understand why it was unreliable on Windows. History of that is captured in the commit messages in case you want to see the sordid details.

Hmm so I understand that I should just ammend commit and run it again as there might be a chance I didnt break it. I was not able to debug it in my Idea for some strange reason, so I cannot tell if this test was affected by my changes

File: pom.xml
Comment (jglick) [inline pom.xml:263]:
```suggestion
```

File: pom.xml
Comment (jglick) [inline pom.xml:273]:
```suggestion
```

Comment (MartinKosicky) [reply pom.xml]: I must say that I wasnt able to compile it locally without these temporary .pom changes, but I assume they will be removed before it will be merged

**Comment by MartinKosicky** [comment]
I did a small refactor of the calculation of the headName. I dont know what is the best practice in this project, I made a dedicated public static method, so that it could be tested standalone. Do you prefer a utility class or can we leave it as is ?

**Comment by MartinKosicky** [comment]
I need it in incrementals, so I can make an integration test in workflow-cps

**Comment by jglick** [comment]
> I need it in incrementals, so I can make an integration test in workflow-cps, so i converted it to to PR (removed draft)

Incrementals are published from draft PRs.

File: pom.xml
Comment (jglick) [inline pom.xml:64]: Update https://github.com/jenkinsci/git-plugin/blob/d7159dc8566217e0e161fb67c8771585ade67f95/pom.xml#L78 accordingly.

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java
Comment (jglick) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:292]: Stylistically I would recommend
```suggestion
            private HeadNameResult(String headName, String prefix) {
```
and then moving `calculateHeadName` inside this nested class, and renaming it accordingly (to simply `calculate`, or `create`, or `make`, or `of`, etc.).

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java
Comment (jglick) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:287]:
```suggestion
        static class HeadNameResult
```
Not used outside this class and the test, right?

(In which case you could also simplify a bit by dropping `private` from the fields and removing the getters—this is just an internal struct, no need for accessors.)

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java
Comment (jglick) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:324]:
```suggestion
                if (branchSpecExpandedName.startsWith(prefix)) {
```

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java
Comment (jglick) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:316]:
```suggestion
            if (branchSpecExpandedName.startsWith(Constants.R_TAGS)){
```

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java
Comment (jglick) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:346]:
```suggestion
```

(gratuitous diff hunk)

File: src/main/java/jenkins/plugins/git/GitSCMFileSystem.java
Comment (jglick) [inline src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:359]:
```suggestion
            if (_build != null) {
```

File: src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java
Comment (jglick) [inline src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:466]:
```suggestion
```

File: src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java
Comment (jglick) [inline src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:435]:
```suggestion
    public void calculate_head_name_with_env() throws Exception {
```

File: src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java
Comment (jglick) [inline src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:439]:
```suggestion
        assertEquals("master", result1.getHeadName());
        assertEquals(Constants.R_HEADS, result1.getPrefix());
```

File: src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java
Comment (jglick) [inline src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:454]:
```suggestion
        assertEquals("${BRANCH}", result4.getHeadName());
        assertEquals(Constants.R_HEADS, result4.getPrefix());
```

File: src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java
Comment (jglick) [inline src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:433]: May also want to assert behavior with a null env, undefined variable, etc.

**Review by jglick — APPROVED** [review]

**Comment by jglick** [comment]
https://github.com/jenkinsci/git-plugin/pull/898#discussion_r448561267 should be investigated. Also that PR touches `GitTelescope` (whatever that is).

**Comment by MartinKosicky** [comment]
I apolagize, I had my linter reconfigured for wrong project, now it should look better

File: src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java
Comment (MartinKosicky) [reply src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:433]: there is one for null env, but this is internal method, and there are many checks already in the calling method

File: pom.xml
Comment (MartinKosicky) [reply pom.xml:64]: Ok i updated it, I hope correctly

**Comment by jglick** [comment]
@MartinKosicky by the way when you force-push inside a PR, it breaks the ability of GitHub to display incremental diffs to reviewers. Generally I advise simply pushing commits to the branch, and leaving it to up the maintainer whether to squash-merge or true-merge at the end. (Some projects rebase-merge and so the list of commits inside the PR branch is actually critical, but this is rare in Jenkins components.)

**Comment by MartinKosicky** [comment]
> ply pushing commits to the branch, and leaving it to up the maintainer whether to squash-merge or true-merge at the end. (Some projects rebase-merge and so the list of commits inside the PR branch is actually critical, but this is rare in Jenkins components.)

I'm sorry, I thought this was the prefered way (I read it somewhere on the internet that it's prefered in github's PRs, as this is my first contribution to a github project). I will do it only incrementally from now. Also I did some ammends because the builder was broken, in that case should I make a dummy commit to retrigger build?

**Comment by MarkEWaite** [comment]
> I advise simply pushing commits to the branch, and leaving it to up the maintainer whether to squash-merge or true-merge at the end.

Considering the number of times that I've made mistakes visible in the history of the master branch of the git plugin, I don't think a history of mistakes and missteps by others will feel at all out of place in the git plugin.  If the submitter asks for a squash merge or if the merge history is especially "long and dirty", then I'll squash merge.

**Comment by jglick** [comment]
> I read it somewhere on the internet that it's prefered in github's PRs

I am afraid every project has its own conventions, and preferences are certainly not consistent inside @jenkinsci which is quite decentralized.

> I did some ammends because the builder was broken, in that case should I make a dummy commit to retrigger build?

That is one way to trigger a fresh build of your own PR, yes. (People with write access to the repository can **Re-run** from the **Checks** tab. Fewer people have permission to directly **Rebuild** or **Replay** on ci.jenkins.io.)

**Comment by MartinKosicky** [comment]
@jglick I would like to ask what is the flow here, for example when the https://github.com/jenkinsci/scm-api-plugin/pull/160 will be ok and merged, should we wait until the scm-api-plugin new version will be released (i dont know what is the time interval here, or release schedule) before marking this PR as ready for review?

**Comment by jglick** [comment]
`scm-api` is set up for CD so release occurs automatically shortly after merging PRs with applicable labels. Ditto `workflow-cps`. I cannot speak for `git`.

**Comment by MarkEWaite** [comment]
> should we wait until the scm-api-plugin new version will be released (i don't know what is the time interval here, or release schedule) before marking this PR as ready for review?

You can mark the PR as ready for review as soon as you feel it is ready.  No need to wait for releases of the dependent plugins so long as your pull request refers to an incremental build of the dependent plugins.  If your pull request refers to incremental builds of the dependent plugins, then I can perform testing with an install of those dependent plugins along with a build of your pull request.

**Comment by MartinKosicky** [comment]
> [#898 (comment)](https://github.com/jenkinsci/git-plugin/pull/898#discussion_r448561267) should be investigated. Also that PR touches `GitTelescope` (whatever that is). Need to examine prior art.

Yes i saw that PR, and that is why I moved expansion at start, and also added tests to make sure that if BRANCH is */master it doesnt get converted to */*/master

**Comment by MartinKosicky** [comment]
fixed some problem with merge conflict

File: pom.xml
Comment (jglick) [inline pom.xml:105]:
```suggestion
      <version>621.vda_a_b_055e58f7</version>  <!-- TODO until in BOM -->
```
https://github.com/jenkinsci/scm-api-plugin/releases/tag/621.vda_a_b_055e58f7

File: pom.xml
Comment (jglick) [inline pom.xml:195]:
```suggestion
      <version>621.vda_a_b_055e58f7</version>  <!-- TODO until in BOM -->
```

**Comment by MarkEWaite** [comment]
> Ok so If I understand correctly, since we change the version from incremental to that which was produced by CI/CD is that correct and can this be merged or are we going to wait for that BOM (which I know about only theoretically) ?

Since the scm-api that is now referenced in the pom is a released version of scm-api, this change could be merged.  No need to wait for scm api to appear in the bom.

The `TODO` is to remind us that once the version is included in the bom, then the `version` entry for scm-api can be removed.

**Comment by MartinKosicky** [comment]
> > Ok so If I understand correctly, since we change the version from incremental to that which was produced by CI/CD is that correct and can this be merged or are we going to wait for that BOM (which I know about only theoretically) ?
> 
> Since the scm-api that is now referenced in the pom is a released version of scm-api, this change could be merged. No need to wait for scm api to appear in the bom.
> 
> The `TODO` is to remind us that once the version is included in the bom, then the `version` entry for scm-api can be removed.

Hmm oki so i guess ill wait for some additional feedback , because i am not able to merge this PR

**Comment by MarkEWaite** [comment]
> Hmm oki so i guess ill wait for some additional feedback , because i am not able to merge this PR

I still need to review the PR.  I have others in my review queue ahead of this PR.  I will be unavailable next week while I take vacation.  It will be at least two weeks before I can review this PR.  If other maintainers would like to review it, they are welcome to do so.

The next release of the master branch of the git plugin will not happen for at least two weeks.  I don't want to risk releasing a new version of the plugin shortly before I disappear for a week of vacation.  It is better for the project if I am available for the week after a git plugin release in case there are surprises or other problems.

**Comment by MartinKosicky** [comment]
> > Hmm oki so i guess ill wait for some additional feedback , because i am not able to merge this PR
> 
> I still need to review the PR. I have others in my review queue ahead of this PR. I will be unavailable next week while I take vacation. It will be at least two weeks before I can review this PR. If other maintainers would like to review it, they are welcome to do so.
> 
> The next release of the master branch of the git plugin will not happen for at least two weeks. I don't want to risk releasing a new version of the plugin shortly before I disappear for a week of vacation. It is better for the project if I am available for the week after a git plugin release in case there are surprises or other problems.

Ah oki then, i dont mean to rush anyone. Oki so I will check in 2 weeks or so, and enjoy vacation:)

**Comment by MartinKosicky** [comment]
Just wanted to let you know that we took the incremental plugin builds for now, and put in our production jenkins, and configured all jobs to use lw checkout now,  it works perfectly

**Comment by MartinKosicky** [comment]
Hello, this is just a friendly reminder as I am not sure if it wasn't forgotten :)

**Comment by MarkEWaite** [comment]
> Hello, this is just a friendly reminder as I am not sure if it wasn't forgotten :)

Thanks for the reminder.  I've returned from a week of vacation, will be in the office for three days, then will be out of the office for four days.  I still have other reviews that precede this review in my queue, so it will be at least 2 weeks before I can review it.

Other maintainers are welcome to review it if they wish.

**Comment by MartinKosicky** [comment]
Hello pinging again as this PR is opened too long, can we either finish it up somehow or probabbly add some improvements?   @MarkEWaite or @dwnusbaum or @car-roll  .

**Comment by MartinKosicky** [comment]
Hmm

**Comment by MartinKosicky** [comment]
Ping again (i apolagize in advance for pinging)

**Comment by MarkEWaite** [comment]
Thanks for the ping.  I'll be busy all this weekend preparing for DevOps World and then busy all week at DevOps World.  I may be able to start the review a week or two after DevOps World

**Review by raul-arabaolaza — APPROVED** [review]

**Comment by MartinKosicky** [comment]
> Thanks for the ping. I'll be busy all this weekend preparing for DevOps World and then busy all week at DevOps World. I may be able to start the review a week or two after DevOps World

Hello is it time now? :)

**Comment by MarkEWaite** [comment]
> > Thanks for the ping. I'll be busy all this weekend preparing for DevOps World and then busy all week at DevOps World. I may be able to start the review a week or two after DevOps World
> 
> Hello is it time now? :)

Not yet.  I'm on vacation with family and friends and won't return until the end of this week.  I hope to start the review early next week, after some restart time as I return.

**Comment by link89** [comment]
Looking forward for this improvement, it suffers when working with a monorepo and not being able to turn on lightweight checkout.

**Comment by MartinKosicky** [comment]
@MarkEWaite Hello can we like do this quick, there are only 2 files in this, one is test and one is the file. I am afraid that the build system will change soon and i will have to start over on this. Can you find some hour at least or 15 minutes to quick check it? some guys already approved it. There are also tests added. Please :)

**Review by MarkEWaite — APPROVED** [review]
Thanks for your patience @MartinKosicky .  Approved..

I need to perform some interactive testing of an incremental build before it is released.  I expect that release will happen later this week.

**Comment by MartinKosicky** [comment]
Just to clarify, who will merge the branch, you after the manual test?

**Comment by MarkEWaite** [comment]
> Just to clarify, who will merge the branch, you after the manual test?

Yes, I'll merge the branch and I'll perform the release.

**Comment by MarkEWaite** [comment]
After building and installing the workflow-cps plugin with the change, I was able to confirm that this resolves [JENKINS-42971](https://issues.jenkins.io/browse/JENKINS-42971).  Thanks!

**Comment by MartinKosicky** [comment]
Thanks :)

**Pull request merged**
